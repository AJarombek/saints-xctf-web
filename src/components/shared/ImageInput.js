/**
 * Container component for an input element that has an identification icon and a validation icon.
 * @author Andrew Jarombek
 * @since 5/2/2020
 */

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const ImageInput = ({ type, name, placeholder, autoComplete = "", maxLength, onChange, icon,
                      status = NONE }) => {

  let statusIcon, statusClass;
  switch (status) {
    case NONE:
      statusIcon = null;
      statusClass = 'none';
      break;
    case SUCCESS:
      statusIcon = '\u0052';
      statusClass = 'success';
      break;
    case WARNING:
      statusIcon = '\u004f';
      statusClass = 'warning';
      break;
    case FAILURE:
      statusIcon = '\u0051';
      statusClass = 'failure';
      break;
    default:
      statusIcon = null;
      statusClass = 'none';
  }

  return (
    <div className="sxctf-image-input">
      {icon ?
        <img src={icon} alt="" />
        :
        <div className="no-icon"> </div>
      }
      <input
        name={name}
        type={type}
        placeholder={placeholder}
        autoComplete={autoComplete}
        maxLength={maxLength}
        onChange={onChange}
      />
      <div>
        <div className={classnames("status", statusClass)}>
          <p>{statusIcon}</p>
        </div>
      </div>
    </div>
  );
};

export const NONE = 0;
export const SUCCESS = 1;
export const WARNING = 2;
export const FAILURE = 3;

ImageInput.Status = {
  NONE, SUCCESS, WARNING, FAILURE
};

ImageInput.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  autoComplete: PropTypes.string,
  maxLength: PropTypes.number,
  onChange: PropTypes.func.isRequired,
  icon: PropTypes.any,
  status: PropTypes.oneOf([NONE, SUCCESS, WARNING, FAILURE])
};

export default ImageInput;
