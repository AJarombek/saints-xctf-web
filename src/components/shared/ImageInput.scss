/*
 * Sass styling for the ImapgeInput component
 * @author Andrew Jarombek
 * @since 5/6/2020
 */

.sxctf-image-input {
  display: flex;
  border: none;
  border-radius: 5px;
  background-color: #fdfdfd;

  height: 50px;
  width: 250px;

  input {
    background-color: transparent;
    border: none;
    width: 200px;
  }

  img {
    width: 36px;
    height: 36px;
    margin: 7px;
  }
}
