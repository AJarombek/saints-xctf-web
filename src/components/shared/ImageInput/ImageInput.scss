/*
 * Sass styling for the ImapgeInput component
 * @author Andrew Jarombek
 * @since 5/6/2020
 */

@import "../../../fonts";
@import "../../../vars";

.sxctf-image-input {
  display: flex;
  border: none;
  border-radius: 5px;
  background-color: #fdfdfd;

  height: 50px;
  width: 250px;

  img {
    flex-shrink: 0;
    width: 36px;
    height: 36px;
    margin: 7px;
  }

  .no-icon {
    flex-shrink: 0;
    width: 20px;
  }

  // Styles for Chrome auto-completion.
  input:-webkit-autofill,
  input:-webkit-autofill:hover,
  input:-webkit-autofill:focus {
    -webkit-text-fill-color: $spot-palette-brown;
    transition: background-color 5000s ease 0s;
  }

  input {
    flex-shrink: 1;
    flex-grow: 1;
    min-width: 100px;

    @include roboto;
    font-size: 12px;
    background-color: transparent;
    border: none;
    width: 200px;

    &:focus {
      outline: none;
    }
  }

  div {
    flex-shrink: 0;

    display: flex;
    align-items: center;
    justify-content: center;
    width: 30px;

    .status {
      display: flex;
      align-items: center;
      justify-content: center;
      width: 20px;
      height: 20px;
      border-radius: 50%;

      &.none {
        background-color: transparent;
      }

      &.success {
        background-color: $status-success;

        p {
          font-size: 22px;
        }
      }

      &.warning {
        background-color: $status-warning;

        p {
          font-size: 24px;
        }
      }

      &.failure {
        background-color: $status-failure;

        p {
          font-size: 24px;
        }
      }

      p {
        @include elegant-icons;
        color: white;
        margin: 0;
      }
    }
  }
}
