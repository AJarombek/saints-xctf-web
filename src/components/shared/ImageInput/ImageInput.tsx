/**
 * Container component for an input element that has an identification icon and a validation icon.
 * @author Andrew Jarombek
 * @since 5/2/2020
 */

import React, { ChangeEvent, KeyboardEvent, useRef } from 'react';
import classNames from 'classnames';
import { ClassValue } from 'classnames/types';

enum ImageInputStatus {
  NONE,
  SUCCESS,
  WARNING,
  FAILURE
}

interface Props {
  type: string;
  name: string;
  placeholder: string;
  minValue?: number;
  autoComplete?: string;
  maxLength?: number;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => any;
  onKeyUp?: (event: KeyboardEvent<HTMLInputElement>) => any;
  useCustomValue?: boolean;
  value?: string;
  icon?: any;
  status: ImageInputStatus;
  className?: ClassValue;
}

const ImageInput: React.FunctionComponent<Props> = ({
  type,
  name,
  placeholder,
  minValue,
  autoComplete = '',
  maxLength,
  onChange,
  onKeyUp,
  useCustomValue,
  value,
  icon,
  status = ImageInputStatus.NONE,
  className
}) => {
  const inputRef = useRef(null);

  let statusIcon, statusClass;
  switch (status) {
    case ImageInputStatus.NONE:
      statusIcon = null;
      statusClass = 'none';
      break;
    case ImageInputStatus.SUCCESS:
      statusIcon = '\u0052';
      statusClass = 'success';
      break;
    case ImageInputStatus.WARNING:
      statusIcon = '\u004f';
      statusClass = 'warning';
      break;
    case ImageInputStatus.FAILURE:
      statusIcon = '\u0051';
      statusClass = 'failure';
      break;
    default:
      statusIcon = null;
      statusClass = 'none';
  }

  return (
    <div className={classNames('sxctf-image-input', className)}>
      {icon ? <img src={icon} alt="" /> : <div className="no-icon"> </div>}
      <input
        name={name}
        type={type}
        placeholder={placeholder}
        min={minValue}
        value={useCustomValue ? value : inputRef.current?.value}
        autoComplete={autoComplete}
        maxLength={maxLength}
        onChange={onChange}
        onKeyUp={onKeyUp}
        ref={inputRef}
      />
      <div>
        <div className={classNames('status', statusClass)}>
          <p>{statusIcon}</p>
        </div>
      </div>
    </div>
  );
};

export { ImageInputStatus };
export default ImageInput;
