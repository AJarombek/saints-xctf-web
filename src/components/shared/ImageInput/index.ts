import ImageInput, { ImageInputStatus } from './ImageInput';

export { ImageInputStatus };
export default ImageInput;
