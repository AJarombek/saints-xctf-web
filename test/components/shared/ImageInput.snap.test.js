/**
 * Snapshot test for the ImageInput component.
 * @author Andrew Jarombek
 * @since 8/16/2020
 */

import React from 'react';
import ImageInput, {ImageInputStatus} from '../../../src/components/shared/ImageInput';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <ImageInput
      type="text"
      name="test1"
      placeholder="Test 1"
      onChange={() => {}}
      status={ImageInputStatus.NONE}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
