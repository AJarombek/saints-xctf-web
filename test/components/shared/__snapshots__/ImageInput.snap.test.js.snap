// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sxctf-image-input"
  data-cypress="image-input-test1"
>
  <div
    className="no-icon"
  >
     
  </div>
  <input
    autoComplete=""
    name="test1"
    onChange={[Function]}
    placeholder="Test 1"
    type="text"
  />
  <div>
    <div
      className="status none"
    >
      <p />
    </div>
  </div>
</div>
`;
